src/Features/Core/Portable/InheritanceMargin/AbstractInheritanceMarginService.cs

```diff
@@ -0,0 +1,124 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.FindUsages;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+using Microsoft.CodeAnalysis.SymbolMapping;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.InheritanceMargin
+{
+    using SymbolKeyAndLineNumberArray = ImmutableArray<(SymbolKey, int lineNumber)>;
+
+    internal abstract partial class AbstractInheritanceMarginService : IInheritanceMarginService
+    {
+        /// <summary>
+        /// Given the syntax nodes to search,
+        /// get all the method, event, property and type declaration syntax nodes.
+        /// </summary>
+        protected abstract ImmutableArray<SyntaxNode> GetMembers(IEnumerable<SyntaxNode> nodesToSearch);
+
+        /// <summary>
+        /// Get the token that represents declaration node.
+        /// e.g. Identifier for method/property/event and this keyword for indexer.
+        /// </summary>
+        protected abstract SyntaxToken GetDeclarationToken(SyntaxNode declarationNode);
+
+        protected abstract string GlobalImportsTitle { get; }
+
+        public async ValueTask<ImmutableArray<InheritanceMarginItem>> GetInheritanceMemberItemsAsync(
+            Document document,
+            TextSpan spanToSearch,
+            CancellationToken cancellationToken)
+        {
+            var (project, symbolKeyAndLineNumbers) = await GetMemberSymbolKeysAsync(document, spanToSearch, cancellationToken).ConfigureAwait(false);
+
+            // if we didn't remap the symbol to another project (e.g. remapping from a metadata-as-source symbol back to
+            // the originating project), then we're in teh same project and we should try to get global import
+            // information to display.
+            var includeGlobalImports = project == document.Project;
+
+            if (!includeGlobalImports && symbolKeyAndLineNumbers.IsEmpty)
+                return ImmutableArray<InheritanceMarginItem>.Empty;
+
+            return await GetInheritanceMemberItemAsync(
+                project,
+                documentForGlobalImports: includeGlobalImports ? document : null,
+                spanToSearch,
+                symbolKeyAndLineNumbers,
+                cancellationToken).ConfigureAwait(false);
+        }
+
+        private async ValueTask<(Project remapped, SymbolKeyAndLineNumberArray symbolKeyAndLineNumbers)> GetMemberSymbolKeysAsync(
+            Document document,
+            TextSpan spanToSearch,
+            CancellationToken cancellationToken)
```

Comment:
this was an extract method. however, isntead of calling into oop, it determines what data it wants to call into oop with.